File: rails-server-manager/app/models/project.rb
class Project < ApplicationRecord
  # Associations
  belongs_to :owner, class_name: 'User'
  belongs_to :user_group, optional: true
  has_many :servers, dependent: :destroy

  # Validations
  validates :name, presence: true
  validates :owner, presence: true

  # Scopes
  scope :for_user, ->(user) {
    where(owner: user)
      .or(where(user_group_id: user.user_group_ids))
      .distinct
  }

  # Methods
  def accessible_by?(user)
    user.can_access_project?(self)
  end

  def servers_count
    servers.count
  end

  def servers_status_summary
    {
      online: servers.where(status: 'online').count,
      offline: servers.where(status: 'offline').count,
      unknown: servers.where(status: 'unknown').count
    }
  end
end

File: laravel-server-manager/app/Models/Project.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Project extends Model
{
    use HasFactory;

    protected $fillable = ['name', 'description', 'owner_id', 'user_group_id'];

    // Relationships
    public function owner()
    {
        return $this->belongsTo(User::class, 'owner_id');
    }

    public function userGroup()
    {
        return $this->belongsTo(UserGroup::class);
    }

    public function servers()
    {
        return $this->hasMany(Server::class);
    }

    // Helper methods
    public function serversStatusSummary()
    {
        return [
            'online' => $this->servers()->where('status', 'online')->count(),
            'offline' => $this->servers()->where('status', 'offline')->count(),
            'unknown' => $this->servers()->where('status', 'unknown')->count(),
        ];
    }
}
